### ISSUE TITLE
Do not panic on zero-byte packet

### ISSUE THREAD

**Issue by rzezeski** [oxidecomputer/opte#197]
I hit this fun panic today, where a zero-byte packet is being sent by `ddmd` (I'm guess that's the DDM daemon). While it is kind of silly that our network stack allows a zero-byte packet to make it this far we should account for this in OPTE. `Packet::wrap` should probably return a `Result` or `Option`, and we should make use of the configurable panic flag from #194 to either panic when this situation occurs or just fire a probe and bump a stat.

```
> ::status
debugging crash dump /var/crash/kalm/vmcore.45 (64-bit) from kalm
operating system: 5.11 helios-1.0.21180 (i86pc)
build version: remotes/ci-build/xde-0-gd365ee6a2b

image uuid: dbfaa3c3-7f11-6f0f-d6b9-a6d558e1a44b
panic message: panicked at 'no bytes found', /home/rpz/oxidecomputer/opte/opte/src/engine/packet.rs:1230:13
dump content: kernel pages only
> 
> $C
fffffe007c6f84a0 vpanic()
fffffe007c6f8520 rust_begin_unwind+0x8a()
fffffe007c6f8560 _ZN4core9panicking9panic_fmt17h7d7971299bdd0756E+0x2d()
fffffe007c6f8600 _ZN4opte6engine6packet47Packet$LT$opte..engine..packet..Initialized$GT$4wrap17h7db70314098cdb88E+0x220()
fffffe007c6f9130 xde_rx+0x43()
fffffe007c6f9180 mac_promisc_dispatch_one+0xac(fffffe58dae50388, fffffe5870718740, 0, 0)
fffffe007c6f9200 mac_promisc_dispatch+0x82(fffffe5896944010, fffffe5870718740, fffffe5895402878, 0)
fffffe007c6f9250 mac_provider_tx+0x57(fffffe5896944010, 0, fffffe5870718740, fffffe5895402878)
fffffe007c6f92e0 mac_bcast_send+0x28a(fffffe5899ab7b98, fffffe5895402878, fffffe5870718740, 1)
fffffe007c6f9390 mac_tx_send+0x15a(fffffe5895402878, 0, fffffe5870718740, fffffe007c6f93b8)
fffffe007c6f9440 mac_tx_single_ring_mode+0x82(fffffe5899c6eb80, fffffe5870718740, 0, 0, 0)
fffffe007c6f9500 mac_tx+0x1a9(fffffe5895402878, fffffe5870718740, 0, 0, 0)
fffffe007c6f9550 str_mdata_fastpath_put+0x8e(fffffe5899bfd5f8, fffffe5870718740, 0, 0)
fffffe007c6f95a0 dld_wput+0x1ac(fffffe58c7db78e0, fffffe5870718740)
fffffe007c6f9610 putnext+0x233(fffffe58c7db7650, fffffe5870718740)
fffffe007c6f9720 ip_xmit+0x632(fffffe5870718740, fffffe5893dcd798, 806000, 30, cb2e54d3, 0, fffffe5800000000, fffffe594e03a530)
fffffe007c6f97b0 ip_postfrag_loopcheck+0x9c(fffffe5870718740, fffffe5893dcd798, 806000, 30, cb2e54d3, 0, ffffffff00000000, fffffe594e03a530)
fffffe007c6f9a00 ire_send_wire_v6+0x126(fffffe5899c63010, fffffe5870718740, fffffe59212df810, fffffe594e03a3c0, fffffe5890d21ef0)
fffffe007c6f9a90 ire_send_multicast_v6+0xa6(fffffe5899c63010, fffffe5870718740, fffffe59212df810, fffffe594e03a3c0, fffffe5890d21ef0)
fffffe007c6f9b10 conn_ip_output+0x1d4(fffffe5870718740, fffffe594e03a3c0)
fffffe007c6f9c20 icmp_output_newdst+0x644(fffffe5972a69b00, fffffe5870718740, 0, fffffe58f8f2dde8, fffffe58968d8188, 2bb, fffffe594e03a3c0)
fffffe007c6f9cb0 rawip_send+0x4b7(fffffe5972a69b00, fffffe5870718740, fffffe007c6f9e50, fffffe58968d8188)
fffffe007c6f9d40 so_sendmsg+0x24a(fffffe5935c91550, fffffe007c6f9e50, fffffe007c6f9e80, fffffe58968d8188)
fffffe007c6f9da0 socket_sendmsg+0x62(fffffe5935c91550, fffffe007c6f9e50, fffffe007c6f9e80, fffffe58968d8188)
fffffe007c6f9e40 sendit+0x1ab(d, fffffe007c6f9e50, fffffe007c6f9e80, 8000)
fffffe007c6f9f00 sendto+0x88(d, 1e53eb8, 8, 8000, fffffc7fea7ff308, 20)
fffffe007c6f9f10 sys_syscall+0x17d()
> 
> fffffe5870718740::mblk
            ADDR FL TYPE    LEN   BLEN              RPTR             DBLK
fffffe5870718740 0  data    0     80    fffffe59212df7f0 fffffe59212df780
> 
> fffffe5870718740::print -t mblk_t
mblk_t {
    struct msgb *b_next = 0
    struct msgb *b_prev = 0
    struct msgb *b_cont = 0
    unsigned char *b_rptr = 0xfffffe59212df7f0
    unsigned char *b_wptr = 0xfffffe59212df7f0
    struct datab *b_datap = 0xfffffe59212df780
    unsigned char b_band = 0
    unsigned char b_tag = 0
    unsigned short b_flag = 0
    queue_t *b_queue = 0

> ::stacks -m xde -t panic
THREAD           STATE    SOBJ                COUNT
fffffe5899edec40 PANIC    <NONE>                  1
                 0xfffffe5936302a10
                 0xfffffffffb8a6a31
                 rust_begin_unwind+0x8a
                 _ZN4core9panicking9panic_fmt17h7d7971299bdd0756E+0x2d
                 _ZN4opte6engine6packet47Packet$LT$opte..engine..packet..Initialized$GT$4wrap17h7db70314098cdb88E+0x220
                 xde_rx+0x43
                 mac_promisc_dispatch_one+0xac
                 mac_promisc_dispatch+0x82
                 mac_provider_tx+0x57
                 mac_bcast_send+0x28a
                 mac_tx_send+0x15a
                 mac_tx_single_ring_mode+0x82
                 mac_tx+0x1a9
                 str_mdata_fastpath_put+0x8e
                 dld_wput+0x1ac
                 putnext+0x233
                 ip_xmit+0x632
                 ip_postfrag_loopcheck+0x9c
                 ire_send_wire_v6+0x126
                 ire_send_multicast_v6+0xa6
                 conn_ip_output+0x1d4
                 icmp_output_newdst+0x644
                 rawip_send+0x4b7
                 so_sendmsg+0x24a
                 socket_sendmsg+0x62
                 sendit+0x1ab
                 sendto+0x88

> fffffe5899edec40::thread -p
            ADDR             PROC              LWP             CRED
fffffe5899edec40 fffffe5899f47000 fffffe58969d4840 fffffe58968d8188

> fffffe5899f47000::ps
S PID    PPID   PGID   SID    UID    FLAGS      ADDR             NAME
R 699    698    9      9      0      0x4a004000 fffffe5899f47000 ddmd
```

**Comment by rzezeski** [comment]
In case someone picks this up before me: I would suggest we go with a `Result` return and change all panic calls in the `wrap()` function to some discriminant on an error type. We should change both `Packet<Initialized>` and `Packet<Uninitialized>`. Note that in the case of an `Uninitialized` packet we **expect** that there are no bytes written, in the case of `Initialized` we expect the opposite: at least 1 byte written. Though that's sure to fail to parse, so I suppose we could also expect some minimum of say at least a 14 byte ethernet header. Anyways, the main thrust is let's not panic.

**Comment by rzezeski** [comment]
Addressed in #305 (17048766fadac81403e4283e4b1dec460d019d39).
